keps/sig-auth/3766-referential-authorization/README.md

```diff
@@ -413,80 +429,121 @@ rejected".
+    // pattern(s).
+    Subject Subject       `json:"subject"`
+
+    // From refers to the group and resource that these references originate from.
+    From    GroupResource `json:"from"`
+
+    // To refers to the group and resource that these references target.
+    To      GroupResource `json:"to"`
+
+    // For refers to the purpose of this reference. (Cluster)ReferenceGrants
+    // matching the From, To, and For of this resource will be authorized for
+    // the Subject of this resource.
+    //
+    // This value must be a valid DNS label as defined per RFC-1035.
+    For     string        `json:"for"`
+}
 
-	// Spec defines the desired state of ReferenceGrant.
-	Spec ReferenceGrantSpec `json:"spec,omitempty"`
+// Subject is a copy of RBAC Subject that excludes APIGroup.
+type Subject struct {
+    // Kind of object being referenced. Values defined by this API group are
+    // "User", "Group", and "ServiceAccount". If the Authorizer does not
+    // recognized the kind value, the Authorizer should report an error.
+    Kind string `json:"kind"`
+    // Name of the object being referenced.
+    Name string `json:"name"`
+    // Namespace of the referenced object.  If the object kind is non-namespace,
+    // such as "User" or "Group", and this value is not empty the Authorizer
+    // should report an error. +optional
+    Namespace string `json:"namespace,omitempty"`
+}
+```
+
+#### ClusterReferenceGrant
 
-	// Note that `Status` sub-resource has been excluded at the
-	// moment as it was difficult to work out the design.
-	// A `Status` sub-resource may be added in the future.
+```golang
+// ClusterReferenceGrant authorizes same-namespace references matching the
+// specified pattern.
+type ClusterReferenceGrant struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // From refers to the group, resource, and pthat that these references
+    // originate from.
+    From    GroupVersionResourcePath `json:"from"`
+
+    // To refers to the group and resource that these references target.
+    To      GroupResource `json:"to"`
+
+    // For refers to the purpose of this reference. Subjects of
+    // ClusterReferenceConsumers will be authorized to follow references matching
+    // the From, To, and For of this resource.
+    //
+    // This value must be a valid DNS label as defined per RFC-1035.
+    For     string        `json:"for"`
 }
 
-// +kubebuilder:object:root=true
-// ReferenceGrantList contains a list of ReferenceGrant.
-type ReferenceGrantList struct {
-	metav1.TypeMeta `json:",inline"`
-	metav1.ListMeta `json:"metadata,omitempty"`
-	Items           []ReferenceGrant `json:"items"`
+type GroupVersionResourcePath struct {
+    Group    string `json:"group"`
+    Resource string `json:"resource"`
+    Version  string `json:"version"`
+    // JSONPath is the JSONPath which this reference may come from.
+    JSONPath string `json:"jsonPath"`
```

Comment:
```suggestion
    // JSONPath is the JSONPath which this reference may come from.
    //+optional
    JSONPath string `json:"jsonPath,omitempty"`
```

The JSONPath should be optional, right?

Reply:
I can't think of any way to implement this without a jsonPath (or CEL path). This all relies on the authorizer being able to follow the path of a reference to authorize access to the resource that is targeted by the reference that this jsonPath points to.

Reply:
Ok, thanks for the clarification 👍